File: src/main/java/io/sch/historyscan/domain/contexts/analysis/clocrevisions/RevisionScore.java
package io.sch.historyscan.domain.contexts.analysis.clocrevisions;

import java.math.BigDecimal;

import static java.util.Comparator.comparing;
import static java.util.Comparator.reverseOrder;

public record RevisionScore(
        long score) implements Comparable<RevisionScore> {

    public static RevisionScore of(int revisions, long currentNbLines) {
        if (currentNbLines == 0) {
            currentNbLines = 1;
        }
        var revisionsScore = calculateRevisionsScore(revisions, currentNbLines);
        return new RevisionScore(revisionsScore);
    }

    static long calculateRevisionsScore(int numberOfRevisions, long nbLines) {
        return BigDecimal.valueOf(numberOfRevisions)
                .multiply(BigDecimal.valueOf(nbLines))
                .longValue();
    }

    @Override
    public int compareTo(RevisionScore o) {
        return comparing(RevisionScore::score, reverseOrder())
                .compare(this, o);
    }
}

File: historyscan-back/src/main/kotlin/io/sch/historyscan/domain/contexts/analysis/clocrevisions/RevisionScore.kt
package io.sch.historyscan.domain.contexts.analysis.clocrevisions

import java.math.BigDecimal

class RevisionScore(val score: Long) : Comparable<RevisionScore> {

    override fun compareTo(other: RevisionScore): Int {
        return Comparator.comparing(RevisionScore::score, Comparator.reverseOrder())
            .compare(this, other)
    }

    companion object {
        fun of(revisions: Int, currentNbLines: Long): RevisionScore {
            var nblines = 0
            if (currentNbLines == 0L) {
                nblines = 1
            }
            val revisionsScore = calculateRevisionsScore(revisions, nblines)
            return RevisionScore(revisionsScore)
        }

        @JvmStatic
        fun calculateRevisionsScore(numberOfRevisions: Int, nbLines: Int): Long {
            return BigDecimal.valueOf(numberOfRevisions.toLong())
                .multiply(BigDecimal(nbLines))
                .toLong()
        }
    }
}
